aggsender/validator/validate_certificate.go

```diff
@@ -0,0 +1,262 @@
+	if err != nil {
+		return fmt.Errorf("error checking first certificate because can't unmarshal metadata. Err: %w", err)
+	}
+	if metadataUnmarshal.FromBlock != 1 {
+		// The first certificate must start from block 0
+		return fmt.Errorf("first certificate must start from block 1, but got: %d",
+			metadataUnmarshal.FromBlock)
+	}
+	if params.Certificate.Height != 0 {
+		// The first certificate must have height 0
+		return fmt.Errorf("first certificate must have height 0, but got: %d",
+			params.Certificate.Height)
+	}
+	return nil
+}
+
+// GetCertificatePreBuildParams prepares the parameters needed to build a certificate based
+// on incomming certificate
+func (a *CertificateValidator) GetCertificatePreBuildParams(ctx context.Context,
+	params VerifyIncommingRequests) (*types.CertificatePreBuildParams, error) {
+	if params.Certificate == nil {
+		return nil, fmt.Errorf("preBuildParams. Err: %w", ErrNilCertificate)
+	}
+	lastSentCertificate, err := AgglayerCertificateHeaderToAggsender(params.PreviousCertificate)
+	if err != nil {
+		return nil, fmt.Errorf("preBuildParams. failed to convert previous certificate to Aggsender format: %w", err)
+	}
+	metadataUnmarshal, err := types.NewCertificateMetadataFromHash(params.Certificate.Metadata)
+	if err != nil {
+		return nil, fmt.Errorf("preBuildParams. Error unmarshal cert.metadata. Err: %w", err)
+	}
+	blockRange, err := metadataUnmarshal.BlockRange()
+	if err != nil {
+		return nil, fmt.Errorf("preBuildParams. failed to get block range from certificate metadata: %w", err)
+	}
+
+	l1InfoRoot, err := a.l1InfoTreeDataQuerier.GetL1InfoRootByLeafIndex(ctx, params.Certificate.L1InfoTreeLeafCount-1)
+	if err != nil {
+		return nil, fmt.Errorf("preBuildParams. failed to get L1 Info tree root by leaf count %d: %w",
+			params.Certificate.L1InfoTreeLeafCount, err)
+	}
+
+	return &types.CertificatePreBuildParams{
+		BlockRange:          blockRange,
+		RetryCount:          0, // TODO: ???
+		CertificateType:     guessCertificateType(params.Certificate, metadataUnmarshal.CertificateType()),
+		LastSentCertificate: lastSentCertificate,
+		L1InfoTreeWhichToProve: &types.CertificateL1InfoTree{
+			L1InfoTreeRootFromWhichToProve: l1InfoRoot.Hash,
+			L1InfoTreeLeafCount:            params.Certificate.L1InfoTreeLeafCount,
+		},
+		CreatedAt: metadataUnmarshal.CreatedAt,
+	}, nil
+}
+
+// guessCertificateType tries to guess the certificate type based on the certificate and metadata.
+func guessCertificateType(certificate *agglayertypes.Certificate,
+	metadataCertType types.CertificateType) types.CertificateType {
+	if metadataCertType != types.CertificateTypeUnknown {
+		return metadataCertType
+	}
+	// Metadata doesn't have the cert type,  I will try to guess from the certificate
+	// TODO: Double check this logic... what about optimistic, PP have something in this field
+	if certificate.AggchainData != nil {
+		proof, err := certificate.AggchainData.MarshalJSON()
```

Comment:
You don't have to call the `MarshalJSON` function. You can just try to to convert it:

```go
proof, ok := certificate.AggchainData.(*agglayertypes.AggchainDataProof)
if ok {
	return types.CertificateTypeFEP
}

return types.CertificateTypePP
```

This way we don't have to do serialization to guess it.